### DISCUSSION TITLE
Show/hide system tray with clicking

### DISCUSSION THREAD

**Discussion by anhnamtran** [polybar/polybar#2954] [Show and tell]
With the new system tray implementation, I wondered if there is a way to show/hide the system tray on click to keep things tidy.

There is! Combining different features including [triggering actions](https://polybar.readthedocs.io/en/latest/user/actions.html#triggering-actions), [IPC](https://polybar.readthedocs.io/en/latest/user/ipc.html), and sending signals to scripts.

The gist of my implementation is as follows:

**Bar module**
```ini
[bar/mainbar]
enable-ipc = true
; the ordering doesn't matter since tray_toggle is hidden when tray is visible
modules-right = tray_toggle tray
```

**Tray Module**
```ini
[prefs]
; custom preferences for different modules to query
start_tray_hidden = true

[module/tray]
type = internal/tray

; The action format code is to send a signal to the toggling script
format = %{A1:kill -USR1 $(pgrep -f tray_toggle.sh):}>%{A} <tray>

hidden = ${prefs.start_tray_hidden}
```

**Tray toggling module**
```ini
[module/tray_toggle]
type = custom/script

; this lets the underlying script know whether we're starting with the tray visible or not
env-START_TRAY_HIDDEN=${prefs.start_tray_hidden}

exec = $HOME/config/polybar/scripts/tray_toggle.sh
tail = true

click-left = kill -USR1 %pid%
```

**tray_toggle.sh**
```sh
#!/usr/bin/env bash
CACHE_DIR="$HOME/.cache/polybar-scripts"
mkdir -p "$CACHE_DIR"

# If the lock file exists, the tray is currently hidden
LOCK_FILE="tray_toggle.lock"

SLEEP_PID=0

show_tray() {
  polybar-msg action tray module_show &>/dev/null
  echo ""
}

hide_tray() {
  polybar-msg action tray module_hide &>/dev/null
  echo ">"
}

handle_click() {
  if [ -f "$CACHE_DIR/$LOCK_FILE" ]; then
    show_tray
    rm -f "$CACHE_DIR/$LOCK_FILE"
  else
    hide_tray
    touch "$CACHE_DIR/$LOCK_FILE"
  fi
}

handle_start() {
  if [ "$START_TRAY_HIDDEN" = "true" ]; then
    hide_tray
    touch "$CACHE_DIR/$LOCK_FILE"
  else
    show_tray
    rm -f "$CACHE_DIR/$LOCK_FILE"
  fi
}

handle_sig() {
  # Signal handler cleans up sleep process
  if (( $SLEEP_PID != 0 )); then
    kill $SLEEP_PID &>/dev/null
  fi
  handle_click
}

handle_start

trap "handle_sig" SIGUSR1

# Keep running until sleep is killed
while true; do
  sleep infinity &
  SLEEP_PID=$!
  wait
done
```

This then creates the following effect (I changed the icon in the scripts because github couldn't load the one in the GIF):
![polybar_show_hide_tray](https://user-images.githubusercontent.com/12238924/235329261-9237d4b0-06b5-483a-b9d2-7ff8cfbc0331.gif)

I may have overengineered due to the following reasons:
1. I wasn't able to find a way to get the visibility of a module from outside of polybar
  - This prompted the necessity of the cache file
2. I wasn't able to find any similar projects, discussions, or issues relating to showing and hiding a module based on clicking.

I'm posting here to gauge interest and see if anyone might have a better implementation of this.

**Comment by patrick96** [comment]
Very nice solution! Yeah there is no way to get any kind of runtime information out of polybar.

If you don't care about preserving state between restarts, the entire toggle logic could just be a text module with `click-left = #tray.module_toggle`.

Alternatively, if you are willing to drop the initial hidden state functionality. You could always display the arrow in the `tray_toggle` module you could probably get rid of the convoluted (but required) logic for sending signals to the script.
Instead the `tray_toggle` module could also be a text module:

```dosini
[module/tray_toggle]
type = custom/text
label = ">"
click-left = tray_toggle.sh
```

And the `tray_toggle.sh` script only contains the `handle_click` logic.

But, if you both want the preservation of the state (and optionally different text depending on the state) and a configurable start state, I don't see any way to make it better or more compact than what you already have.

**Reply by anhnamtran** [reply]
Thanks for the reply! Glad that I didn't over engineer anything, I'll see about making this its own repo so others might have access to it.

**Reply by rocky0x4e** [reply]
thank @anhnamtran for the idea and @patrick96 for the suggestion, I was able to make a simpler version, click the icon and it shows the tray for 3s then hides it

config:
```
[module/tray]
type = internal/tray
; The action format code is to send a signal to the toggling script
format = <tray>
hidden = true
tray-size = 99
format-padding = 10px

[module/tray_toggle]
type = custom/text
label = ">>"
click-left = ~/.config/polybar/modules/utils/tray_toggle.sh
```

tray_toggle.sh
```
#!/usr/bin/env bash

CACHE_DIR="$HOME/.cache/polybar-scripts"
mkdir -p "$CACHE_DIR"

# If the lock file exists, the tray is currently hidden
LOCK_FILE="${CACHE_DIR}/tray_toggle.lock"

if [[ ! -f $LOCK_FILE ]]; then
    touch $LOCK_FILE
    polybar-msg action tray module_show &>/dev/null
    sleep 5
    polybar-msg action tray module_hide &>/dev/null
    rm "$LOCK_FILE"
fi
```

**Reply by rocky0x4e** [reply]
Hey, I just found out that the polybar IPC supports toggle action.
So we don't even need the tray_toggle.sh file, just the module config is enough

```
[module/tray_toggle]
type = custom/text
label = ">>"
click-left = polybar-msg action tray module_toggle
```
```
polybar 3.6.3-212-g16102c25
```

**Reply by MalyugaSensei** [reply]
I'm using polybar 3.7.1 and the command ```click-left = polybar-msg action tray module_toggle``` didn't work. Try using ```click-left = #tray.module_toggle```

**Reply by og900aero** [reply]
How can it be solved so that the tray is hidden by default and only appears by clicking on the arrow?
Because I  use hidden = true, but doesnt work:

```
[module/tray]
type = internal/tray
format = <tray>
hidden = true
tray-size = 80%
tray-background = ${colors.background}
```
